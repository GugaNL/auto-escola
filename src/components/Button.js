import React from "react"

//Style
import styled from "styled-components"

//Types
import PropTypes from "prop-types"

const getMainColor = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.main
    case ButtonColors.danger:
      return theme.colors.danger.main
    default:
      return "#e0e0e0"
  }
}

const getDarkColor = ({ theme, color }) => {
  switch (color) {
    case ButtonColors.primary:
      return theme.colors.primary.dark
    case ButtonColors.danger:
      return theme.colors.danger.dark
    default:
      return "#5a6268"
  }
}

const getColorText = (props) => props.theme.colors.primary.text

const getLinkText = (props) => {
  if (props.color === ButtonColors.default) {
    return "#757575"
  }

  return getMainColor(props)
}

const ButtonColors = {
  default: "default",
  primary: "primary",
  danger: "danger",
}

const ButtonsVariants = {
  default: "default",
  outlined: "outlined",
  link: "link",
}

const Button = styled.button`
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: ${getMainColor};
  border: 2px solid ${getMainColor};
  color: ${getColorText};

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover:enabled {
    background-color: ${getDarkColor};
    border-color: ${getDarkColor};
  }
`

const ButtonOutlined = styled(Button)`
  background-color: transparent;
  color: ${getMainColor};

  &:hover:enabled {
    background-color: transparent;
    color: ${getDarkColor};
  }
`

const ButtonLinked = styled(Button)`
  background-color: transparent;
  border-color: transparent;
  color: ${getLinkText};

  &:hover:enabled {
    background-color: transparent;
    border-color: transparent;
    color: ${getDarkColor};
  }
`

const ButtonWrapper = (props) => {
  //Necessário criar pois o componente Button é um styled-component, então o storybook não reconhece o props dele
  switch (props.variant) {
    case ButtonsVariants.outlined:
      return <ButtonOutlined {...props} />
    case ButtonsVariants.link:
      return <ButtonLinked {...props} />
    default:
      return <Button {...props} />
  }
}

ButtonWrapper.defaultProps = {
  type: "button",
  children: undefined,
  color: "default",
  variant: "default",
}

ButtonWrapper.propTypes = {
  type: PropTypes.string,
  children: PropTypes.node,
  color: PropTypes.oneOf(Object.values(ButtonColors)), //Como sempre serão 3 valores fixos então pode-se usar um enum ao invés de string
  variant: PropTypes.oneOf(Object.values(ButtonsVariants)),
}

export default ButtonWrapper
