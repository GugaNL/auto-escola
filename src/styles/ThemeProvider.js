import React from "react"

//Theme
import { ThemeProvider as StyledProvider } from "styled-components"

export const ThemeNames = { light: "light", ocean: "ocean" }

const light = {
  colors: {
    primary: {
      main: "#ffc104",
      dark: "#c79100",
      light: "#fff350",
      text: "#212121",
    },
    danger: {
      main: "#ff3d00",
      dark: "#b22a00",
      text: "#212121",
    },
  },
}

const allThemes = {
  light,
  ocean: {
    ...light,
    colors: {
      ...light.colors,
      primary: {
        main: "#2196f3",
        dark: "#1769aa",
        light: "#4dabf5",
      },
    },
  },
}

const ThemeProvider = ({ theme, children }) => (
  <StyledProvider theme={allThemes[theme]}>{children}</StyledProvider>
)

ThemeProvider.defaultProps = {
  theme: "light",
}

export default ThemeProvider
