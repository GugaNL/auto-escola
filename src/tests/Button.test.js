import React from "react"
import Button from "../components/atoms/Button"
import { render, fireEvent } from "@testing-library/react"
import ThemeProvider from "../styles/ThemeProvider"

test("renders a text in button", () => {
  const { getByText } = render(
    <ThemeProvider>
      <Button>Click here</Button>
    </ThemeProvider>
  )

  expect(getByText("Click here")).toBeInTheDocument()
})

test("trigger event on click", () => {
  const handleClick = jest.fn() //Função para mock de evento

  const { getByRole } = render(
    <ThemeProvider>
      <Button onClick={handleClick}></Button>
    </ThemeProvider>
  )

  fireEvent.click(getByRole("button"))

  expect(handleClick).toBeCalled()
})

test.each([["default"], ["primary"], ["danger"]])(
  "renders with %s color",
  (color) => {
    const { asFragment } = render(
      <ThemeProvider>
        <Button color={color} />
      </ThemeProvider>
    )

    expect(asFragment()).toMatchSnapshot()
  }
)

test.each([["default"], ["outlined"], ["link"]])(
  "renders with %s variant",
  (variant) => {
    const { asFragment } = render(
      <ThemeProvider>
        <Button variant={variant} />
      </ThemeProvider>
    )

    expect(asFragment()).toMatchSnapshot()
  }
)

test("renders with variant outlined and primary color", () => {
  const { asFragment } = render(
    <ThemeProvider>
      <Button variant="outlined" color="primary" />
    </ThemeProvider>
  )

  expect(asFragment()).toMatchSnapshot()
})

test("renders with variant link and primary color", () => {
  const { asFragment } = render(
    <ThemeProvider>
      <Button variant="link" color="primary" />
    </ThemeProvider>
  )

  expect(asFragment()).toMatchSnapshot()
})
