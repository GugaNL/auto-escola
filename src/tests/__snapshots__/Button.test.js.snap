// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with danger color 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ff3d00;
  border: 2px solid #ff3d00;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #b22a00;
  border-color: #b22a00;
}

<button
    class="c0"
    color="danger"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with default color 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with default variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with link variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #757575;
  padding: 0;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  border-color: transparent;
  color: #5a6268;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with outlined variant 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #5a6268;
  border-color: #5a6268;
}

.c1 {
  background-color: transparent;
  color: #e0e0e0;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  color: #5a6268;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with primary color 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc104;
  border: 2px solid #ffc104;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

<button
    class="c0"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant link and primary color 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc104;
  border: 2px solid #ffc104;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #ffc104;
  padding: 0;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  border-color: transparent;
  color: #c79100;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`renders with variant outlined and primary color 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  cursor: pointer;
  background-color: #ffc104;
  border: 2px solid #ffc104;
  color: #212121;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:not(:disabled) {
  background-color: #c79100;
  border-color: #c79100;
}

.c1 {
  background-color: transparent;
  color: #ffc104;
}

.c1:hover:not(:disabled) {
  background-color: transparent;
  color: #c79100;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;
